pkg/middleware/auth/middleware.go

```diff
@@ -1,152 +1,205 @@
-		sessionManager:       opts.SessionManager,
+		peer:                 peer,
+		wallet:               cfg.Wallet,
+		sessionManager:       sessionManager,
 		transport:            t,
-		allowUnauthenticated: opts.AllowUnauthenticated,
+		allowUnauthenticated: cfg.AllowUnauthenticated,
 		logger:               middlewareLogger,
 	}, nil
 }
 
-// Handler returns standard http middleware
+// Handler returns an HTTP handler that processes authentication messages.
 func (m *Middleware) Handler(next http.Handler) http.Handler {
-	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
-		recorder := newResponseRecorder(w)
-		if req.Method == http.MethodPost && req.URL.Path == "/.well-known/auth" {
-			err := m.transport.HandleNonGeneralRequest(req, recorder)
-			if err != nil {
-				http.Error(recorder, err.Error(), http.StatusUnauthorized)
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		wrappedWriter := httptransport.WrapResponseWriter(w)
+
+		ctx := context.WithValue(r.Context(), httptransport.RequestKey, r)
+		ctx = context.WithValue(ctx, httptransport.ResponseKey, wrappedWriter)
+		ctx = context.WithValue(ctx, httptransport.NextKey, func() {
+			next.ServeHTTP(wrappedWriter, r)
+		})
+
+		m.logger.Debug("Processing request",
+			slog.String("path", r.URL.Path),
+			slog.String("method", r.Method))
+
+		authMsg, err := httptransport.ParseAuthMessageFromRequest(r)
+		if err != nil {
+			m.logger.Error("Failed to parse auth message", slog.String("error", err.Error()))
+			http.Error(wrappedWriter, err.Error(), http.StatusBadRequest)
+			return
+		}
+
+		if authMsg == nil {
+			if m.allowUnauthenticated {
+				r = r.WithContext(context.WithValue(r.Context(), httptransport.IdentityKey, constants.UnknownParty))
+				next.ServeHTTP(wrappedWriter, r)
+				return
+			} else {
+				http.Error(wrappedWriter, "Authentication required", http.StatusUnauthorized)
+				return
 			}
-			createResponse(recorder)
+		}
+
+		// At this point, authMsg.IdentityKey should always be set by ParseAuthMessageFromRequest
+		// If it's not set, that's an internal error
+		if authMsg.IdentityKey == nil {
+			m.logger.Error("Internal error: ParseAuthMessageFromRequest returned message with nil IdentityKey")
+			http.Error(wrappedWriter, "Internal authentication error", http.StatusInternalServerError)
 			return
 		}
 
-		req, authMsg, err := m.transport.HandleGeneralRequest(req, recorder)
+		callback, err := m.transport.GetRegisteredOnData()
 		if err != nil {
-			http.Error(recorder, err.Error(), http.StatusUnauthorized)
-			createResponse(recorder)
+			m.logger.Error("No message handler registered", slog.String("error", err.Error()))
+			http.Error(wrappedWriter, "Server configuration error", http.StatusInternalServerError)
 			return
 		}
 
-		next.ServeHTTP(recorder, req)
-
-		err = m.transport.HandleResponse(req, recorder, recorder.body.Bytes(), recorder.statusCode, authMsg)
+		err = callback(ctx, authMsg)
 		if err != nil {
```

Comment:
```suggestion
		if err := callback(ctx, authMsg); err != nil {
```
why not this way ? I think is more readable ?